state/logdb/buf.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package logdb
+
+import (
+	"sync"
+	"time"
+
+	"github.com/juju/errors"
+	"github.com/juju/utils/clock"
+
+	"github.com/juju/juju/state"
+)
+
+// Logger provides an interface for writing log records.
+type Logger interface {
+	// Log writes the given log records to the logger's storage.
+	Log([]state.LogRecord) error
+}
+
+// BufferedLogger wraps a Logger, providing a buffer that
+// accumulates log messages, flushing them to the underlying logger
+// when enough messages have been accumulated.
+type BufferedLogger struct {
+	l             Logger
+	clock         clock.Clock
+	flushInterval time.Duration
+
+	mu         sync.Mutex
+	buf        []state.LogRecord
+	flushTimer clock.Timer
+}
+
+// NewBufferedLogger returns a new BufferedLogger, wrapping the given
+// Logger with a buffer of the specified size and flush interval.
+func NewBufferedLogger(
+	l Logger,
+	bufferSize int,
+	flushInterval time.Duration,
+	clock clock.Clock,
+) *BufferedLogger {
+	return &BufferedLogger{
+		l:             l,
+		buf:           make([]state.LogRecord, 0, bufferSize),
+		clock:         clock,
+		flushInterval: flushInterval,
+	}
+}
+
+// Log is part of the Logger interface.
+//
+// BufferedLogger's Log implementation will buffer log records up to
+// the specified capacity and duration; after either of which is exceeded,
+// the records will be flushed to the underlying logger.
+func (b *BufferedLogger) Log(in []state.LogRecord) error {
+	b.mu.Lock()
```

Comment:
I'm curious if the logic around this would be clearer (or more confusing) if Log just sent a message onto a channel for a worker loop to consume, rather than having locks, etc.
Most of Juju code is based around worker loops.